File: backend/src/main/java/com/backend/petplace/domain/place/importer/KcisaClient.java
package com.backend.petplace.domain.place.importer;

import com.backend.petplace.domain.place.dto.KcisaDto;
import com.backend.petplace.global.config.ImportProperties;
import java.net.URI;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@Component
@RequiredArgsConstructor
public class KcisaClient {

  private final ImportProperties props;

  @Qualifier("kcisaWebClient")
  private final WebClient client;

  public List<KcisaDto.Item> fetchPage(int pageNo) {
    URI uri = URI.create(String.format(
        "%s?serviceKey=%s&numOfRows=%d&pageNo=%d",
        props.baseUrl(), props.serviceKey(), props.pageSize(), pageNo
    ));

    KcisaDto root = client.get()
        .uri(uri)
        .retrieve()
        .bodyToMono(KcisaDto.class)
        .block();

    if (root == null || root.response() == null || root.response().body() == null) return List.of();
    var items = root.response().body().items();
    return (items == null || items.item() == null) ? List.of() : items.item();
  }
}

File: backend/src/main/java/com/backend/petplace/domain/place/importer/KcisaClient.kt
package com.backend.petplace.domain.place.importer

import com.backend.petplace.domain.place.dto.KcisaDto
import com.backend.petplace.global.config.ImportProperties
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import java.net.URI

@Component
class KcisaClient(
    private val props: ImportProperties,
    @Qualifier("kcisaWebClient")
    private val client: WebClient
) {

    fun fetchPage(pageNo: Int): List<KcisaDto.Item> {
        val uri = URI.create(
            "${props.baseUrl}?serviceKey=${props.serviceKey}&numOfRows=${props.pageSize}&pageNo=$pageNo"
        )

        val root = client.get()
            .uri(uri)
            .retrieve()
            .bodyToMono(KcisaDto::class.java)
            .block()

        val body = root?.response?.body ?: return emptyList()
        val items = body.items ?: return emptyList()
        return items.item ?: emptyList()
    }
}
